src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs

```diff
@@ -174,29 +176,58 @@ internal MemoryCacheEntryOptions GetMemoryCacheEntryOptions()
         {
             var content = await output.GetChildContentAsync();
 
-            var stringBuilder = new StringBuilder();
-            using (var writer = new StringWriter(stringBuilder))
+            using (var writer = new CharBufferTextWriter())
             {
                 content.WriteTo(writer, HtmlEncoder);
+                return new CharBufferHtmlContent(writer.Buffer);
+            }
+        }
+
+        private class CharBufferTextWriter : TextWriter
+        {
+            public CharBufferTextWriter()
+            {
+                Buffer = new PagedCharBuffer(CharArrayBufferSource.Instance);
+            }
+
+            public override Encoding Encoding => Null.Encoding;
+
+            public PagedCharBuffer Buffer { get; }
+
+            public override void Write(char value)
+            {
+                Buffer.Append(value);
+            }
+
+            public override void Write(char[] buffer, int index, int count)
+            {
+                Buffer.Append(buffer, index, count);
             }
 
-            return new StringBuilderHtmlContent(stringBuilder);
+            public override void Write(string value)
+            {
+                Buffer.Append(value);
+            }
```

Comment:
I'll add in `WriteLine(string)` since it does a bunch of extra work: http://referencesource.microsoft.com/#mscorlib/system/io/textwriter.cs,452